cpp/server/proxy.h

```diff
@@ -0,0 +1,170 @@
+#ifndef CERT_TRANS_SERVER_PROXY_H_
+#define CERT_TRANS_SERVER_PROXY_H_
+
+#include <event2/http.h>
+#include <evhttp.h>
+#include <functional>
+#include <vector>
+
+#include "base/macros.h"
+#include "log/cluster_state_controller.h"
+#include "monitoring/monitoring.h"
+#include "net/url_fetcher.h"
+
+
+namespace util {
+class Task;
+}  // namespace util
+
+
+namespace cert_trans {
+
+namespace {
+static Counter<std::string>* total_proxied_requests(
+    Counter<std::string>::New("total_proxied_requests", "path",
+                              "Number of proxied API requests by path."));
+static Counter<std::string, int>* total_proxied_responses(
+    Counter<std::string, int>::New("total_proxied_responses", "path",
+                                   "status_code",
+                                   "Number of proxied API requests by path "
+                                   "and status code."));
+}  // namespace
+
+
+template <class Logged>
+class ClusterStateController;
+class UrlFetcher;
+
+
+template <class Logged>
+class Proxy {
+ public:
+  Proxy(const ClusterStateController<Logged>* controller, UrlFetcher* fetcher,
```

Comment:
Or... maybe you could take an ```std::function``` to give you what you want form the ```controller```? Since it's just this one thing? With a clever signature, you could just ```bind``` it straight on?